### PR TITLE
Enabled macros for transform plugins

### PR DESCRIPTION
**Pull request by knazir**
The following properties have been macro-enabled:

`CSVFormatter`:
- format
- delimiter

`CSVParser`:
- format

`CloneRecord`:
- copies

`Compressor`:
- compressor

`Decoder`:
- decode

`Decompressor`:
- decompressor

`Decryptor`:
- decryptFields

`Encoder`:
- encode

`Encryptor`:
- encryptFields

`Hasher`:
- hash
- fields

`Normalize`:
- Updated `validate` method to align with other plugins.

`StreamFormatter`: 
- body
- header
- format

`ValueMapper`:
- mapping
- defaults

`XMLParser`:
- inputField
- encoding

`XMLToJSON`:
- outputFields

### CODE REVIEW

**Comment by knazir** [comment]
Guild is breen: http://builds.cask.co/browse/HYP-BAD53-5

**Comment by knazir** [comment]
Guild is breen: http://builds.cask.co/browse/HYP-BAD53-6

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/CSVFormatter.java [hunk 1]
```diff
@@ -224,6 +212,25 @@ public Config(String format, String delimiter, String schema) {
       this.delimiter = delimiter;
       this.schema = schema;
     }
+
+    private void validate() {
+      if (!containsMacro("delimiter") && !delimMap.containsKey(delimiter)) {
+        throw new IllegalArgumentException("Unknown delimiter '" + delimiter + "' specified. ");
```
Comment (albertshau) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/CSVFormatter.java]: ... specified. Allowed values are " + Joiner.on(", ").join(delimMap.keySet()).

**Comment by albertshau** [comment]
minor comment, otherwise lgtm

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java [hunk 2]
```diff
@@ -228,7 +228,7 @@ public NormalizeConfig(String fieldMapping, String fieldNormalizing, String outp
       this.outputSchema = outputSchema;
     }
 
-    void validateConfig() {
+    private void validate() {
       Preconditions.checkArgument(!Strings.isNullOrEmpty(fieldMapping), "Fields to mapped cannot be empty.");
```
Comment (gsps1) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:232]: can `fieldMapping` and `fieldNormalizing` be macro enabled ?

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java [hunk 3]
```diff
@@ -133,6 +134,7 @@ public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter)
 
     @Name("outputField")
     @Description("The field containing the XML string to convert into a JSON string.")
+    @Macro
```
Comment (gsps1) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java:137]: can `inputField` also be macro enabled?

**Comment by gsps1** [comment]
just couple of fields which are missing macro fields. fix them and merge.

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java (hunk 2 above)
Comment (knazir) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:232]: `fieldMapping` and `fieldNormalizing` are used for schema validation and we planned not to support macros on any properties used for this as we can't perform these checks in `prepareRun`, right?

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java (hunk 3 above)
Comment (knazir) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java:137]: Same as above, inputField is used for schema validation, which can't be deferred to `prepareRun` as of now I think

**Comment by knazir** [comment]
@gsps1  replied to comments

**Comment by gsps1** [comment]
LGTM

File: transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java (hunk 3 above)
Comment (gsps1) [inline transform-plugins/src/main/java/co/cask/hydrator/plugin/XMLToJSON.java:137]: oh cool. i thought only outputSchema was used.

**Pull request merged**